Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -17,118 +17,414 @@ extension MongoCollection {
+         *   - upsert: When `true`, creates a new document if no document matches the query
+         */
+        public init(filter: Document, update: Document, arrayFilters: [Document]? = nil, collation: Document? = nil,
+                    upsert: Bool? = nil) {
+            self.filter = filter
+            self.update = update
+            self.options = UpdateModelOptions(arrayFilters: arrayFilters, collation: collation, upsert: upsert)
+        }
+
+        /// Adds the `updateOne` operation to a bulk write
+        public func addToBulkWrite(bulk: BulkWriteOperation, index: Int) throws {
+            let opts = try BsonEncoder().encode(self.options)
+            var error = bson_error_t()
 
-    /// When `true`, creates a new document if no document matches the query.
-    public let upsert: Bool?
+            guard mongoc_bulk_operation_update_one_with_opts(bulk._bulk, filter.data, update.data, opts.data, &error) else {
+                throw MongoError.invalidArgument(message: toErrorString(error))
+            }
+        }
+    }
+
+    /// A model for an `updateMany` operation within a bulk write.
+    public struct UpdateManyModel: WriteModel {
+        private let filter: Document
+        private let update: Document
+        private let options: UpdateModelOptions
+
+        /**
+         * Create a `updateMany` operation for a bulk write.
+         *
+         * - Parameters:
+         *   - filter: A `Document` representing the match criteria
+         *   - update: A `Document` containing update operators
+         *   - arrayFilters: A set of filters specifying to which array elements an update should apply
+         *   - collation: Specifies a collation to use
+         *   - upsert: When `true`, creates a new document if no document matches the query
+         */
+        public init(filter: Document, update: Document, arrayFilters: [Document]? = nil, collation: Document? = nil,
+                    upsert: Bool? = nil) {
+            self.filter = filter
+            self.update = update
+            self.options = UpdateModelOptions(arrayFilters: arrayFilters, collation: collation, upsert: upsert)
+        }
+
+        /// Adds the `updateMany` operation to a bulk write
+        public func addToBulkWrite(bulk: BulkWriteOperation, index: Int) throws {
+            let opts = try BsonEncoder().encode(self.options)
+            var error = bson_error_t()
+
+            guard mongoc_bulk_operation_update_many_with_opts(bulk._bulk, filter.data, update.data, opts.data, &error) else {
+                throw MongoError.invalidArgument(message: toErrorString(error))
+            }
+        }
+    }
 }
 
-/// A model for an `updateMany` command.
-public struct UpdateManyModel: WriteModel {
-    /// A `Document` representing the match criteria.
-    public let filter: Document
+/// A protocol indicating write operations that can be batched together using `MongoCollection.bulkWrite`.
+public protocol WriteModel {
```

Comment:
I was considering was moving this protocol _into_ MongoCollection, alongside other models that implement it. Would that just make things more tedious for users since they might have to declare an array of `MongoCollection<Document>.WriteModel` types? If so, we can leave as-is.

Reply:
Ah, I just tried this out and it turns out protocols cannot be nested within other declarations. See https://github.com/apple/swift/blob/master/test/decl/nested/protocol.swift

I suppose that settles it. We can keep this as-is :)

Reply:
huh, TIL. that makes it easy.